### PR TITLE
metric: add started and killed walredo processes counter

### LINKED ISSUE #5722: metric: launched and killed walredo processes
In OOM situations, knowing exactly how many walredo processes there were at a time would help afterwards to understand why was pageserver OOM killed.

We need two counters:
- walredo processes started
- walredo processes shutdown

Good first issue:

- pageserver/src/walredo.rs is the walredo implementation
- We could increment the started counter near launching the process
- We could increment the killed counter in the process waiting newtype wrapper
- metric could be `pageserver_walredo_processes`, an IntCounterVec with "operation" key
  - the two concrete counters could be `operation="started"` and `operation="shutdown"`
  - the "shutdown" separate from "killed" if we'd ever want to distinguish on the ones we need to kill in addition to finding them having exited

### PR DESCRIPTION
**Pull request by rmodpur**
In OOM situations, knowing exactly how many walredo processes there were at a time would help afterwards to understand why was pageserver OOM killed. Add `pageserver_wal_redo_process_total` metric to keep track of total wal redo process started, shutdown and killed since pageserver start.

Closes #5722

### CODE REVIEW

**Comment by rmodpur** [comment]
@koivunej could you please review

**Review by koivunej — CHANGES_REQUESTED** [review]
I think this is looking good. Could I ask you to instead of a single counter_vec create a type hosting the 3 different counters and make that type the `pub(crate) static WAL_REDO_PROCESS_COUNTERS: Lazy<WalRedoProcessCounters> = ...`?

Rationale: walredo is very critical for us, and even though the contention from these counter accesses should always be low, it would allow us moving the strings ("started", "killed", "shutdown") to one place -- the `WalRedoProcessCounters::default` where they are created once and can be accessed by field after that.

File: pageserver/src/metrics.rs [hunk 1]
```diff
@@ -1261,6 +1262,31 @@ pub(crate) static WAL_REDO_RECORD_COUNTER: Lazy<IntCounter> = Lazy::new(|| {
     .unwrap()
 });
 
+pub(crate) struct WalRedoProcessCounters {
+    pub(crate) started: GenericCounter<metrics::core::AtomicU64>,
+    pub(crate) shutdown: GenericCounter<metrics::core::AtomicU64>,
+    pub(crate) killed: GenericCounter<metrics::core::AtomicU64>,
```
Comment (koivunej) [inline pageserver/src/metrics.rs:1268]:
```suggestion
    pub(crate) started: IntCounter,
    pub(crate) shutdown: IntCounter,
    pub(crate) killed: IntCounter,
```

**Review by koivunej — APPROVED** [review]
This is great! Thanks, could you look at the suggestion before I approve the CI run?

File: pageserver/src/metrics.rs (hunk 1 above)
Comment (rmodpur) [reply pageserver/src/metrics.rs]: done

**Comment by rmodpur** [comment]
@koivunej sorry but had to fix a clippy lint

**Comment by koivunej** [comment]
Zero worries, thanks for handling it!

**Review by koivunej — APPROVED** [review]
Aah it was the `|| ...())` suggestion, I should had caught that. It always bites me as well!

**Review by problame — CHANGES_REQUESTED** [review]
I'm not a fan of shoving the counters under the same metric.

Labels are supposed to add context / enricht the event.

But these labels are really different event types.

I'd argue for two separate counters:

* `..._starts` (no label)
* `..._stops` with label `cause` and values: `problem_during_launch` or `drop`

**Comment by koivunej** [comment]
> I'm not a fan of shoving the counters under the same metric.
> 
> Labels are supposed to add context / enricht the event.
> 
> But these labels are really different event types.

@rmodpur please wait before making changes. The PR is now very close to as it was asked in the issue. The overall design of the metrics names and how to use them should be handled separatedly.

The PR is great work anyways, thanks, and we'll take it from here.

**Review by problame — APPROVED** [review]

**Pull request merged**